### PR TITLE
Compensate for non-minimal UTF-8 encodings

### LINKED ISSUE #3320: Writing random output to console output handle fails with no last error set
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.19002.1000]

Any other software?
WSL2
```

# Steps to reproduce
https://github.com/microsoft/WSL/issues/4495
1. Install Ubuntu 18.04 distro from store
2. `for i in /bin/*; do head ${i}; done`
3. Observe wsl window hang or crash

# Expected behavior

Contents of the files are displayed, no hang / crash.

# Actual behavior

WSL window will close, or hang.  Seems like an issue with (or around) Utf8ToWideCharParser::_InvolvedParse.

### PR DESCRIPTION
**Pull request by miniksa**
## Summary of the Pull Request
- Permits substitution of Unicode Replacement for non-minimal codepoint encodings in UTF-8.

## PR Checklist
* [x] Closes #3320
* [x] I work here
* [x] Tests added/passed
* [x] Comments added, follow on issue filed #3378 
* [x] I'm a core contributor

## Detailed Description of the Pull Request / Additional comments
- We do a lot of work to figure out whether or not we have some invalid UTF-8 inside our own internal parser. We're correctly identifying in the first full-conversion pass that something is amiss with the stream of text we were given. And inside the second pass for involved parsing, we are identifying and removing obviously wrong sequences like those that have a lead byte without the correct number of continuations, continuations that come from nowhere, and so on. But there's one big gap:

- We're not correctly identifying non-minimal forms of characters. Specifically, what is causing the crash, is non-minimal representations of the null character U+0000. The minimal form of this character in UTF-8 is 0x00. But technically, it could also be written as any of the following:
1. 0xC0 0x80 - Lead byte of 2 byte sequence and a single continuation. 
2. 0xE0 0x80 0x80 - Lead byte of 3 byte sequences and two continuations.
3. 0xF0 0x80 0x80 0x80 - Lead byte of 4 byte sequences and three continuations.
All of which didn't fill any of the payload bits.

- The OS does identify these as invalid non-minimal forms when the data is sent to `MultiByteToWideChar` after we believe we've removed all invalid data and then it errors out because we set the `MB_ERR_INVALID_CHARS` flag.

- If we remove the flag, the error goes away and the OS will substitute one or more U+FFFDs for these sequences and continue past them. 

- This is inconsistent with the rest of our invalid behavior (where we just eat the invalid bytes and walk on instead of substituting them) but the OS doesn't offer that provision as an option.

- We also can't straight up just call the OS in all cases because we want to be available for the case where a caller sends us part of a valid sequence at the end of the buffer and then continues with the next valid pieces in the next call. That is, think of the `putc` case where someone drops 0xe3, 0x81, and 0x99 on three calls in a row. We want to form those together into the correct U+3059 once the third one comes in. Just using `MultiByteToWideChar` straight up will convert each of them into their own U+FFFD on the three calls. Not OK. So we must have *some* knowledge of UTF-8 to allow this valid scenario to happen.

- The solution here is to let the somewhat inconsistent behavior of "replacements for non-minimal sequences but suppress clearly invalid things" to happen. We're doing this because the fix is needed in the Windows product for 20H1, which is today subject to ever-tightening requirements to prepare to ship. The smallest and least risky fix possible is preferable right now.

- #3378 is filed as a follow on to investigate reconciling the somewhat inconsistent behavior as well as other things noted during this investigation in the future to be consumed into Terminal and whatever Windows release comes after 20H1.

## Validation Steps Performed

Ran the repro steps in #3320 on 20h1 previews with WSL2. No longer crashes after (because it no longer returns the error).

Added automated test in utf8ToWideCharParserTests.cpp to ensure that non-minimal forms don't cause `.Parse` to throw an error and turn into some number of replacements instead

### CODE REVIEW

File: src/host/ut_host/Utf8ToWideCharParserTests.cpp [hunk 1]
```diff
@@ -264,6 +264,54 @@ class Utf8ToWideCharParserTests
...
+            0xe0, 0x80, 0x80, // U+0000 as a 3-byte sequence (non-minimal)
+            0x18, 0x77, 0x40, 0x31, // more single byte points
+            0xf0, 0x80, 0x80, 0x80, // U+0000 as a 4-byte sequence (non-minimal)
+            0x59, 0x1f, 0x68, 0x20 // more single byte points
+        };
+
+        // Expected conversion
+        const wchar_t wideData[] = {
+            0x0060, 0x0012, 0x0000, 0x007f,
+            0xfffd, 0xfffd,
```
Comment (DHowett-MSFT) [inline src/host/ut_host/Utf8ToWideCharParserTests.cpp]: reminder: comments to say it's not load-bearing.

File: src/host/ut_host/Utf8ToWideCharParserTests.cpp [hunk 2]
```diff
@@ -264,6 +264,54 @@ class Utf8ToWideCharParserTests
...
+        unsigned int generated = 0;
+        unique_ptr<wchar_t[]> output{ nullptr };
+        auto parser = Utf8ToWideCharParser{ utf8CodePage };
+
+        VERIFY_SUCCEEDED(parser.Parse(data, count, consumed, output, generated));
+        VERIFY_ARE_EQUAL(count, consumed);
+        VERIFY_ARE_EQUAL(wideCount, generated);
+        VERIFY_IS_NOT_NULL(output.get());
+
+        for (int i = 0; i < wideCount; i++)
```
Comment (DHowett-MSFT) [inline src/host/ut_host/Utf8ToWideCharParserTests.cpp]: nit: this will make the output hard to parse, as it won't actually tell you which index it failed one.

You could make temporary wstrings_view out of both of these, skip the `for` loop and just `VERIFY_ARE_EQUAL` the views... which would proceed to print out the raw strings, but at least it'd give you a bit more info? idk.

File: src/host/utf8ToWideCharParser.cpp [hunk 3]
```diff
@@ -377,8 +377,16 @@ unsigned int Utf8ToWideCharParser::_InvolvedParse(_In_reads_(cb) const byte* con
...
         return 0;
     }
+
+    // By this point, all obviously invalid sequences have been removed.
+    // But non-minimal forms of sequences might still exist.
+    // MB2WC will fail non-minimal forms with MB_ERR_INVALID_CHARS at this point.
+    // So we call with flags = 0 such that non-minimal forms get the U+FFFD
+    // replacement character treatment.
+    // This issue and related concerns are fully captured in future work item GH#3378
+    // for future cleanup and reconciliation.
```
Comment (zadjii-msft) [inline src/host/utf8ToWideCharParser.cpp]: I'd maybe add a reference back to this issue, so that we can recall why this is load-bearing

File: src/host/ut_host/Utf8ToWideCharParserTests.cpp (hunk 2 above)
Comment (miniksa) [reply src/host/ut_host/Utf8ToWideCharParserTests.cpp]: It was copy/pasta from above. But yeah, you're right, a string comparison might be better. WEX::Common::Strings work well for comparison (when viewed in the log)

File: src/host/ut_host/Utf8ToWideCharParserTests.cpp (hunk 1 above)
Comment (miniksa) [reply src/host/ut_host/Utf8ToWideCharParserTests.cpp]: yep thanks.

File: src/host/ut_host/Utf8ToWideCharParserTests.cpp (hunk 2 above)
Comment (miniksa) [reply src/host/ut_host/Utf8ToWideCharParserTests.cpp]: Yep, replaced with WEX::Common::Strings

File: src/host/utf8ToWideCharParser.cpp (hunk 3 above)
Comment (miniksa) [reply src/host/utf8ToWideCharParser.cpp]: done.

**Comment by miniksa** [comment]
> This _seems_ fine to me, esp. based on the analysis in the PR body. How does this treat the utf-test file in #3147?

I believe this resolves #3147 too. But I'm not seeing the pre-change behavior the same way that @egmontkob was in that filing.

I tried catting/typing it out in various forms in conhost and WT (pwsh, cmd, ubuntu, etc.). Before the change, sometimes it gets stuck or incomplete, but not always. And after the change, it seems to be 100% reliable. But I'd prefer if @egmontkob double checked it with the next version to see if the fix satisfies his expectations before closing it.
